File: sdk/typescript/src/orbit.ts
import type { JsonValue, OrbitOptions, ThreadOptions } from "./protocol.js";
import { Thread } from "./thread.js";

const REQUIRED_ENV_VARS = ["CODEX_API_KEY"];

/**
 * The top-level Orbit client. Wraps the `orbit` CLI and is used to start and
 * resume conversation {@link Thread}s.
 */
export class Orbit {
  readonly options: OrbitOptions;
  readonly command: string;

  constructor(options: OrbitOptions = {}) {
    this.options = options;
    this.command = options.command ?? "orbit";
  }

  /** Start a fresh conversation thread. */
  startThread(threadOptions: ThreadOptions = {}): Thread {
    return new Thread(this, threadOptions, undefined);
  }

  /** Reconstruct a thread from a previously persisted session id. */
  resumeThread(sessionId: string, threadOptions: ThreadOptions = {}): Thread {
    return new Thread(this, threadOptions, sessionId);
  }

  /**
   * Build the environment for a spawned CLI. Starts from `process.env`, applies
   * the user-provided `env`, then injects any required variables (such as
   * `CODEX_API_KEY`) from the ambient environment if still missing.
   */
  buildEnv(extra?: Record<string, string>): NodeJS.ProcessEnv {
    const base: NodeJS.ProcessEnv = { ...process.env };
    if (this.options.env) {
      Object.assign(base, this.options.env);
    }
    if (extra) {
      Object.assign(base, extra);
    }
    for (const varName of REQUIRED_ENV_VARS) {
      if (!base[varName] && process.env[varName]) {
        base[varName] = process.env[varName];
      }
    }
    return base;
  }
}

export type { JsonValue };

File: sdk/rust/src/orbit.rs
//! The top-level [`Orbit`] client.

use std::collections::HashMap;
use std::sync::Arc;

use crate::protocol::{OrbitOptions, ThreadOptions};
use crate::thread::Thread;

const REQUIRED_ENV_VARS: &[&str] = &["CODEX_API_KEY"];

/// The top-level Orbit client. Wraps the `orbit` CLI and is used to start and
/// resume conversation [`Thread`]s.
#[derive(Debug, Clone)]
pub struct Orbit {
  pub options: OrbitOptions,
  pub command: String,
}

impl Orbit {
  /// Create a client from the given options.
  pub fn new(options: OrbitOptions) -> Self {
    let command = options
      .command
      .clone()
      .unwrap_or_else(|| "orbit".to_string());
    Self { options, command }
  }

  /// Start a fresh conversation thread.
  pub fn start_thread(&self, options: ThreadOptions) -> Thread {
    Thread::new(Arc::new(self.clone()), options, None)
  }

  /// Reconstruct a thread from a previously persisted session id.
  pub fn resume_thread(&self, session_id: String, options: ThreadOptions) -> Thread {
    Thread::new(Arc::new(self.clone()), options, Some(session_id))
  }

  /// Build the environment for a spawned CLI. Starts from the ambient
  /// environment, applies the user-provided `env`, then injects any required
  /// variables (such as `CODEX_API_KEY`) if still missing.
  pub fn build_env(&self) -> HashMap<String, String> {
    let mut env: HashMap<String, String> = std::env::vars().collect();
    if let Some(user_env) = &self.options.env {
      for (key, value) in user_env {
        env.insert(key.clone(), value.clone());
      }
    }
    for var in REQUIRED_ENV_VARS {
      if !env.contains_key(*var) {
        if let Ok(value) = std::env::var(var) {
          env.insert((*var).to_string(), value);
        }
      }
    }
    env
  }
}
